AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,840 @@ class AzureIoTHub {
+            }
         }
 
-        function _sendEvent(message, done) {
-            // create transfer and send
-            _transfers.event <- _senders.event.createtransfer(message.createAMQPMessage());
-            _log("send event transfer created")
-            _transfers.event.sendasync(function() {
-                _log("in send event transfer callback");
-                if (done) done(null);
-            }.bindenv(this));
+        function _handleTwinGetResponse(reqId, status, parsedMsg) {
+            local cb = _twinRetrievedCb[1];
+            _twinRetrievedCb = null;
+            // If no pending requests, cancel the timer
+            if (_twinPendingRequests.len() == 0) {
+                imp.cancelwakeup(_processQueueTimer);
+                _processQueueTimer = null;
+            }
+            if (!_statusIsOk(status)) {
+                cb(status, null, null);
+                return;
+            }
+            local repProps = null;
+            local desProps = null;
+            try {
+                repProps = parsedMsg["reported"];
+                desProps = parsedMsg["desired"];
+            } catch (e) {
+                _logError("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            cb(0, repProps, desProps);
         }
 
-        function _handleDeliveries(deliveries, cb) {
-            while (deliveries.len()) {
-                local item = AzureIoTHub.Delivery(deliveries.remove(0));
-                cb(null, item);
+        function _handleDirMethodMsg(message, topic) {
+            local methodName = null;
+            local reqId = null;
+            local params = null;
+
+            try {
+                methodName = split(topic, "/")[3];
+                reqId = split(topic, "=")[1];
+                if (message != null && message != "") {
+                    params = http.jsondecode(message);
+                }
+            } catch (e) {
+                _logError("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
             }
+
+            local resp = _onMethodCb(methodName, params);
+            local topic = _topics.dMethodResp + format("%i/?$rid=%s", resp._status, reqId);
+
+            local mqttMsg = _mqttclient.createmessage(topic, http.jsonencode(resp._body), _options.qos);
+
+            local msgSentCb = function (err) {
+                if (err != 0) {
+                    _logError("Could not send Direct Method response. Return code = " + err);
+                }
+            }.bindenv(this);
+
+            mqttMsg.sendasync(msgSentCb);
         }
 
-        function _log(logMsg) {
-            if (_debug) server.log(logMsg);
+        function _processQueue() {
+            local cb = null;
+            local timestamp = null;
+            local isPending = _twinRetrievedCb != null && _twinRetrievedCb[2] != null;
```

Comment:
please either add a check for the `_twinRetrievedCb` length or a comment saying that it's length is "guaranteed" to not be less than 3. Otherwise looks scary

Reply:
Done